addons/pos_sale/tests/test_pos_sale_report.py

```diff
@@ -115,3 +117,37 @@ def test_warehouse(self):
 
         reports = self.env['sale.report'].sudo().search([('product_id', '=', self.product0.id)], order='id', limit=2)
         self.assertEqual(reports[0].warehouse_id.id, self.config.picking_type_id.warehouse_id.id)
+
+    def test_qty_deliverd_qty_to_deliver_in_sales_report(self):
+        """
+            Track the quantity of products ordered based on their picking state. for example : If an order is created for 3 products
+            with the option to ship later, the products will be listed under qty_to_deliver in the sales report until the picking state
+            is validated. Once validated and marked as done, the quantity will shift to qty_delivered.
+        """
+        self.config.ship_later = True
+        self.open_new_session()
+        session = self.pos_session
+
+        orders = []
+
+        orders.append(self.create_ui_order_data([(self.product0, 5)], self.partner_1))
+        orders[0]['shipping_date'] = fields.Date.to_string(fields.Date.today())
+
+        order = self.env['pos.order'].sync_from_ui(orders)
+        order = self.env['pos.order'].browse(order['pos.order'][0]['id'])
+
+        session.action_pos_session_closing_control()
+
+        report = self.env['sale.report'].sudo().search([('product_id', '=', self.product0.id)], order='id')
+
+        self.assertEqual(report.qty_to_deliver, 5)
+        self.assertEqual(report.qty_delivered, 0)
+
+        order.picking_ids.move_ids.quantity = 5.0
+        order.picking_ids.button_validate()
+        # open form and then close to refresh a page to update/refresh sale report
+        stock_picking_form = Form(order.picking_ids)
+        stock_picking_form.save()
```

Comment:
```suggestion
        # flush computations and clear the cache before checking again the report
        self.env.flush_all()
        self.env.clear()
        
        report = self.env['sale.report'].sudo().search([('product_id', '=', self.product0.id)], order='id')
```

We might need to be more explicit with the intent so instead of using the form to "flush" and "clear", let's explicitly call the methods.

Reply:
@caburj Changes are done please check
Thanks!